base/cvd/cuttlefish/host/libs/web/oauth2_consent.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+ * Copyright (C) 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "host/libs/web/oauth2_consent.h"
+
+#include <android-base/file.h>
+#include <unistd.h>
+
+#include <zlib.h>
+
+#include <iostream>
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include "common/libs/fs/shared_buf.h"
+#include "common/libs/fs/shared_fd.h"
+#include "common/libs/utils/contains.h"
+#include "common/libs/utils/json.h"
+#include "common/libs/utils/subprocess.h"
+#include "host/libs/web/http_client/http_client.h"
+#include "host/libs/xdg/directories.h"
+
+namespace cuttlefish {
+namespace {
+
+using android::base::Join;
+using android::base::Tokenize;
+
+Result<std::string> CodeFromUrl(std::string_view url) {
+  static constexpr std::string_view kCodeEq = "code=";
+  std::size_t code_begin = url.find(kCodeEq);
+  CF_EXPECTF(code_begin != std::string_view::npos, "No '{}'", kCodeEq);
+
+  std::size_t code_end = url.find("&", code_begin + 2);
```

Comment:
in `return std::string(url.substr(code_begin + kCodeEq.size(), code_end));`, `code_end` ends up holding the code size, aiming for `code_end` to continue holding the index value suggest the following:

```
code_begin += kCodeEq.size()
std::size_t code_end = url.find("&", code_begin);
if (code_end == std::string::npos) {
  code_end = url.size();
}
return std::string(url.substr(code_begin, code_end - code_begin));
```

Reply:
Rewrote this to remove pieces of the string incrementally, since I'm afraid of making mistakes with a more complex `substr `call.